File: src/Git.cs
using System.Diagnostics;
using System.Text;

namespace GitStateWidget;

public sealed record GitResult(int ExitCode, string StdOut, string StdErr) {
    public bool Ok => ExitCode == 0;
    public IEnumerable<string> Lines => StdOut.Split('\n', StringSplitOptions.RemoveEmptyEntries).Select(l => l.TrimEnd('\r'));
}

/// <summary>
/// Runs git with the process created by us, not by a shell: CreateNoWindow plus redirected
/// streams means neither git nor anything it spawns (git-remote-https, the credential manager)
/// ever gets a console to flash. The environment also forbids every interactive prompt, so an
/// expired credential surfaces as a failed fetch in the panel rather than as a dialog.
/// </summary>
public static class Git {
    public static async Task<GitResult> RunAsync(string repoPath, CancellationToken ct, params string[] args) {
        var psi = new ProcessStartInfo("git") {
            WorkingDirectory = repoPath,
            CreateNoWindow = true,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            StandardOutputEncoding = Encoding.UTF8,
            StandardErrorEncoding = Encoding.UTF8
        };
        foreach (var a in args) {
            psi.ArgumentList.Add(a);
        }
        psi.Environment["GIT_TERMINAL_PROMPT"] = "0";
        psi.Environment["GCM_INTERACTIVE"] = "never";
        psi.Environment["GIT_SSH_COMMAND"] = "ssh -oBatchMode=yes";
        psi.Environment["LC_ALL"] = "C";

        using var p = new Process { StartInfo = psi };
        try {
            p.Start();
        } catch (Exception ex) {
            return new GitResult(-1, "", "could not start git: " + ex.Message);
        }

        var stdout = p.StandardOutput.ReadToEndAsync(ct);
        var stderr = p.StandardError.ReadToEndAsync(ct);
        try {
            await p.WaitForExitAsync(ct);
        } catch (OperationCanceledException) {
            try {
                p.Kill(entireProcessTree: true);
            } catch {
            }
            return new GitResult(-2, "", "timed out");
        }
        return new GitResult(p.ExitCode, await stdout, await stderr);
    }
}

File: src-tauri/src/git.rs
//! Runs git with a process we create ourselves: no console window, streams captured, and every
//! interactive prompt forbidden. An expired credential therefore surfaces as a failed fetch in
//! the panel rather than as a dialog, and nothing git spawns can flash a window.

use std::path::Path;
use std::process::Stdio;
use std::time::Duration;
use tokio::process::Command;

pub struct GitResult {
    pub exit_code: i32,
    pub stdout: String,
    pub stderr: String,
}

impl GitResult {
    pub fn ok(&self) -> bool {
        self.exit_code == 0
    }

    pub fn lines(&self) -> impl Iterator<Item = &str> {
        self.stdout.lines().map(|l| l.trim_end_matches('\r')).filter(|l| !l.is_empty())
    }

    fn failed(msg: impl Into<String>) -> Self {
        Self { exit_code: -1, stdout: String::new(), stderr: msg.into() }
    }
}

pub async fn run(repo: &Path, args: &[&str]) -> GitResult {
    let mut cmd = Command::new("git");
    cmd.args(args)
        .current_dir(repo)
        .env("GIT_TERMINAL_PROMPT", "0")
        .env("GCM_INTERACTIVE", "never")
        .env("GIT_SSH_COMMAND", "ssh -oBatchMode=yes")
        .env("LC_ALL", "C")
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true);

    #[cfg(windows)]
    {
        const CREATE_NO_WINDOW: u32 = 0x0800_0000;
        cmd.creation_flags(CREATE_NO_WINDOW);
    }

    let child = match cmd.spawn() {
        Ok(c) => c,
        Err(e) => return GitResult::failed(format!("could not start git: {e}")),
    };

    match tokio::time::timeout(Duration::from_secs(60), child.wait_with_output()).await {
        Ok(Ok(out)) => GitResult {
            exit_code: out.status.code().unwrap_or(-1),
            stdout: String::from_utf8_lossy(&out.stdout).into_owned(),
            stderr: String::from_utf8_lossy(&out.stderr).into_owned(),
        },
        Ok(Err(e)) => GitResult::failed(e.to_string()),
        Err(_) => GitResult::failed("timed out"),
    }
}
